expression/builtin.go

```diff
@@ -83,10 +121,209 @@ func (b *baseBuiltinFunc) getCtx() context.Context {
+	res, isKindNull, err := b.self.evalDecimal(row)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	if !isKindNull {
+		d.SetMysqlDecimal(res)
+	}
+	return
+}
+
+// evalDecimal will always be overridden.
+func (b *baseDecimalBuiltinFunc) evalDecimal(row []types.Datum) (*types.MyDecimal, bool, error) {
+	return b.self.evalDecimal(row)
+}
+
+func (b *baseDecimalBuiltinFunc) evalInt(row []types.Datum) (int64, bool, error) {
+	val, isKindNull, err := b.self.evalDecimal(row)
+	if err != nil || isKindNull {
+		return 0, isKindNull, errors.Trace(err)
+	}
+	res, err := val.ToInt()
+	return res, false, errors.Trace(err)
+}
+
+func (b *baseDecimalBuiltinFunc) evalReal(row []types.Datum) (float64, bool, error) {
+	val, isKindNull, err := b.self.evalDecimal(row)
+	if err != nil || isKindNull {
+		return 0, isKindNull, errors.Trace(err)
+	}
+	res, err := val.ToFloat64()
+	return res, false, errors.Trace(err)
+}
+
+func (b *baseDecimalBuiltinFunc) evalString(row []types.Datum) (string, bool, error) {
+	val, isKindNull, err := b.self.evalDecimal(row)
+	if err != nil || isKindNull {
+		return "", isKindNull, errors.Trace(err)
+	}
+	return string(val.ToString()), false, errors.Trace(err)
+}
+
+// baseStringBuiltinFunc represents the functions which return string values.
+type baseStringBuiltinFunc struct {
+	baseBuiltinFunc
+}
+
+func (b *baseStringBuiltinFunc) eval(row []types.Datum) (d types.Datum, err error) {
+	val, isKindNull, err := b.self.evalString(row)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	if !isKindNull {
+		d.SetString(val)
+	}
+	return
+}
+
+// evalString will always be overridden.
+func (b *baseStringBuiltinFunc) evalString(row []types.Datum) (string, bool, error) {
+	return b.self.evalString(row)
+}
+
+func (b *baseStringBuiltinFunc) evalInt(row []types.Datum) (int64, bool, error) {
+	val, isKindNull, err := b.self.evalString(row)
+	if err != nil || isKindNull {
+		return 0, isKindNull, errors.Trace(err)
+	}
+	res, err := strconv.ParseInt(val, 10, 64)
+	return res, false, errors.Trace(err)
+}
+
+func (b *baseStringBuiltinFunc) evalReal(row []types.Datum) (float64, bool, error) {
+	val, isKindNull, err := b.self.evalString(row)
+	if err != nil || isKindNull {
+		return 0, isKindNull, errors.Trace(err)
+	}
+	res, err := strconv.ParseFloat(val, 64)
+	return res, false, errors.Trace(err)
+}
+
+func (b *baseStringBuiltinFunc) evalDecimal(row []types.Datum) (*types.MyDecimal, bool, error) {
+	val, isKindNull, err := b.self.evalString(row)
+	if err != nil || isKindNull {
+		return nil, isKindNull, errors.Trace(err)
+	}
+	res := new(types.MyDecimal)
+	err = res.FromString([]byte(val))
+	return res, false, errors.Trace(err)
+}
+
 // builtinFunc stands for a particular function signature.
 type builtinFunc interface {
 	// eval does evaluation by the given row.
 	eval([]types.Datum) (types.Datum, error)
+	// evalInt evaluate int result of builtinFunc by given row.
+	evalInt(row []types.Datum) (int64, bool, error)
```

Comment:
The returned values should be commented or named like
`evalInt(row []types.Datum) (val int64, isNull bool, err error)